lib/Sema/SemaBounds.cpp

```diff
@@ -345,3 +345,86 @@ void Sema::CheckTopLevelBoundsDecls(VarDecl *D) {
   if (!D->isLocalVarDeclOrParm())
     CheckBoundsDeclarations(*this).TraverseVarDecl(D);
 }
+
+
+namespace {
+  class NonModifiyingExprSema : public RecursiveASTVisitor<NonModifiyingExprSema> {
+
+  private:
+    // Represents which kind of modifying expression we have found
+    enum ModifyingExprKind {
+      MEK_Assign,
+      MEK_Increment,
+      MEK_Decrement,
+      MEK_Call,
+      MEK_Comma,
+      MEK_Volatile
+    };
+
+  public:
+    NonModifiyingExprSema(Sema &S) : S(S) {}
+
+    // Assignments are of course modifying
+    bool VisitBinAssign(BinaryOperator* E) {
+      addError(E, MEK_Assign);
+      return false;
+    }
+
+    // Assignments are of course modifying
+    bool VisitCompoundAssignOperator(CompoundAssignOperator *E) {
+      addError(E, MEK_Assign);
+      return false;
+    }
+
+    // Pre-increment/decrement, Post-increment/decrement
+    bool VisitUnaryOperator(UnaryOperator *E) {
+      if (E->isIncrementDecrementOp()) {
+        addError(E,
+          E->isIncrementOp() ? MEK_Increment : MEK_Decrement);
+        return false;
+      }
+
+      return true;
+    }
+
+    // References to volatile variables
+    bool VisitDeclRefExpr(DeclRefExpr *E) {
+      QualType RefType = E->getType();
+      if (RefType.isVolatileQualified()) {
+        addError(E, MEK_Volatile);
+        return false;
+      }
+
+      return true;
+    }
+
+    // Function Calls are defined as modifying
+    bool VisitCallExpr(CallExpr *E) {
+      addError(E, MEK_Call);
+      return false;
+    }
+
+    // Comma Expressions are defined as modifying
+    bool VisitBinComma(BinaryOperator *E) {
+      addError(E, MEK_Comma);
```

Comment:
It is OK to allow comma operators within non-modifying expressions.    The restriction on ',' is a syntactic restriction.  It can't appear at the top-level in a non-modifying expression because non-modifying expressions may appear in comma-separated parameter lists.   Plain C expressions have this problem too, so there is a special syntax for this (an extra production it the grammar, from what I recall).